### PR TITLE
Add a battle calculator performance test [preview]

### PR DESCRIPTION
**Pull request by DanVanAtta**
please do not merge - just a preview to discuss what a performance test could look like.

### CODE REVIEW

**Comment by RoiEXLab** [comment]
@DanVanAtta The build is failing due to checkstyle violations

**Comment by DanVanAtta** [comment]
@RoiEXLab this is a *very* early preview, really want to start conversation. This update is not intended to be mergeable just yet.

**Comment by DanVanAtta** [comment]
We need to discuss in specific how good/reasonable of a strategy this is. Take a chunk of code, run it 50 times and verify latency is under a SLA could lead to non-determinism in the build. This non-determinism may not be that realistic though, in which case we are enforcing the SLA. This would help mitigate the cost of PRs when we ask and check "did this negatively impact battle calc performance?"; so that is quite valuable - is the cost of non-determinism and SLA maintenance greater than that benefit?

I don't think we have the full data to know how often we'll have non-determinism, but I personally suspect it will not be an issue and so we'll have more benefit than cost. I'd like to know everyone else's opinion so we don't rip out SLA tests the first time they fail.

**Comment by ssoloff** [comment]
> Take a chunk of code, run it 50 times and verify latency is under a SLA could lead to non-determinism in the build.

Might be an opportunity to investigate [Travis build stages](https://docs.travis-ci.com/user/build-stages/).  We could have a separate build stage for performance tests that, upon failure, would not affect the release stage.  A performance stage failure is "advisory only" in the sense that artifacts would still be produced, but we may choose not to promote a pre-release to stable due to performance issues.

**Comment by ron-murhammer** [comment]
@DanVanAtta Yeah, we need to be careful about non-determinism around performance testing. I kind of like @ssoloff's idea about some sort of secondary testing stage where we can see the information but it doesn't fail the build.

File: src/integ_test/java/org/triplea/performance/PerfTester.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+import com.google.common.base.Preconditions;
+
+public class PerfTester {
+
+  private PerfTester(final Runnable runnable, final int runCount, final int maxRuntimeMillis) {
+
+    final long start = System.nanoTime();
+
+    IntegerSequence.range(1, runCount).forEach(i -> runnable.run());
```
Comment (RoiEXLab) [inline src/integ_test/java/org/triplea/performance/PerfTester.java]: You should probably use IntStream instead.

**Comment by DanVanAtta** [comment]
> However using mocks we could already verify the original GameData object is only called once at which implies it has been cloned before further usage.

@RoiEXLab interesting suggestion. Part of the trouble is that the game data clone is not fully dominant of the performance. Analyzing calc performance, it looked like there were three actions that were particularly slow, clone was only one of the three. I was not able to pinpoint the exact source of the other two, the rabbit hole went pretty deep at that point.

------------------

@all I think the yoda principle may apply here, if we do not enforce the SLA, then we do not enforce it. So having an optional build I think really will make it optional. I'm concerned then what happens when we slow drift towards that SLA and then maybe breach it. At that point reverting the last change is not really the answer when there have been gradual performance regressions all along.

The real answer for performance is in design, not low level optimization tweaks. We do not want prevent a regression on battle calc performance; I wonder if we should instead focus first on simplifying battle clc so that long term we can get to a clean design that is fast by nature. "Easier to optimize correct code than it is to correct optimized code"

**Pull request closed without merging**